data-prepper-plugins/user-agent-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/useragent/CaffeineCachingParser.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.useragent;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import ua_parser.Client;
+import ua_parser.Device;
+import ua_parser.OS;
+import ua_parser.Parser;
+import ua_parser.UserAgent;
+
+import java.util.function.Function;
+
+/**
+ * A superclass of {@link Parser} which uses Caffeine as a cache.
+ */
+class CaffeineCachingParser extends Parser {
+    private final Cache<String, Client> clientCache;
+    private final Cache<String, UserAgent> userAgentCache;
+    private final Cache<String, Device> deviceCache;
+    private final Cache<String, OS> osCache;
+
+    /**
+     * Constructs a new instance with a given cache size.
+     *
+     * @param cacheSize The size of the cache as a count of items.
+     */
```

Comment:
This is misleading. It creates four caches of the given size. Is this intended? It will use more memory because of this.

Reply:
This is the current behavior of the `CachingParser`. I followed it for consistency.

https://github.com/ua-parser/uap-java/blob/61ffd25de944bbe634ff5180104c9cdc07faad29/src/main/java/ua_parser/CachingParser.java#L30-L33

I did update the Javadoc to explain more.